### PR TITLE
Add functional tests to verify cloud-init nodata is picked up by VM

### PR DESCRIPTION
**Pull request by davidvossel**
Tests now verify that "found datasource (nocloud, local)" is present in the console output in order to verify cloud-init is processed.

### CODE REVIEW

**Comment by davidvossel** [comment]
fixes #426

File: pkg/cloud-init/cloud-init.go [hunk 1]
```diff
@@ -311,7 +311,7 @@ func ApplyMetadata(vm *v1.VM) {
 			return
 		}
 		// TODO Put local-hostname in MetaData once we get pod DNS working with VMs
-		msg := fmt.Sprintf("instance-id: %s.%s\n", domain, namespace)
+		msg := fmt.Sprintf("{ \"instance-id\": \"%s.%s\" }\n", domain, namespace)
```
Comment (davidvossel) [inline pkg/cloud-init/cloud-init.go]: I had to change this because cirros requires that the meta-data be in json format instead of yaml. Cirros reports parsing errors otherwise and won't process the user-data.

Once we expand upon this structure and add more info, we may want to consider a struct that is encoded into json rather than manually formatting the json string as I've done here. I didn't see a reason to add that additional logic for the single 'instance-id' field.

File: tests/vm_userdata_test.go [hunk 2]
```diff
@@ -42,9 +46,16 @@ var _ = Describe("CloudInit UserData", func() {
...
 	tests.PanicOnError(err)
 
-	BeforeEach(func() {
-		tests.BeforeTestCleanup()
-	})
+	Dial := func(vm string, console string) *websocket.Conn {
+		wsUrl, err := url.Parse(flag.Lookup("master").Value.String())
+		Expect(err).ToNot(HaveOccurred())
+		wsUrl.Scheme = "ws"
+		wsUrl.Path = "/apis/kubevirt.io/v1alpha1/namespaces/" + tests.NamespaceTestDefault + "/vms/" + vm + "/console"
```
Comment (rmohr) [inline tests/vm_userdata_test.go]: Ca you use

```go
url, err := kubecli.NewVirtHandlerClient(virtClient).ForNode("master").ConsoleURI(vm)
url.Scheme = "ws"
```
to construct the URL, still not optimal, but at least we don't distribute the URL everywhere in our code. Also in console_test.go where this was copied from.

**Comment by davidvossel** [comment]
> Could you add a a payload to the test which actually does something? 
yup, the user-data prints something the console when it is executed right now. I can just do another check to verify that line is present in the console output.

File: tests/vm_userdata_test.go (hunk 2 above)
Comment (davidvossel) [reply tests/vm_userdata_test.go]: yep, i'll take a look at that.

Comment (davidvossel) [reply tests/vm_userdata_test.go]: What you suggest isn't equivalent to what's currently being done here.

ConsoleURI(vm) returns a url directly to virt-handler.  What we're doing in the console tests and this userdata test is contacting the api server.

**Comment by davidvossel** [comment]
@rmohr The updated patch verifies that the vm actually executes the user-data script. It turns out that's really simple to verify.   Whatever I echo out from the user-data script makes it to the console output. I just set a magicStr to echo out of the user-data and verify the console output gets it.

File: tests/utils.go [hunk 3]
```diff
@@ -511,13 +511,11 @@ func NewRandomVMWithEphemeralDisk(containerImage string) *v1.VM {
...
 
-func NewRandomVMWithUserData(containerImage string, cloudInitDataSource string) (*v1.VM, error) {
-
+func NewRandomVMWithUserData(cloudInitDataSource string) (*v1.VM, error) {
 	switch cloudInitDataSource {
 	case "noCloud":
-		userData := "#cloud-config\npassword: atomic\nssh_pwauth: True\nchpasswd: { expire: False }\n"
-
-		vm := NewRandomVMWithEphemeralDisk(containerImage)
+		userData := "#!/bin/sh\n\necho 'printed from cloud-init userdata'\n"
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply tests/utils.go]: fixed

File: tests/vm_userdata_test.go (hunk 2 above)
Comment (rmohr) [reply tests/vm_userdata_test.go]: True.

File: tests/vm_userdata_test.go [hunk 4]
```diff
@@ -56,71 +70,71 @@ var _ = Describe("CloudInit UserData", func() {
...
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+	})
 
+	Context("CloudInit Data Source NoCloud", func() {
+		It("should launch vm with cloud-init data source NoCloud", func(done Done) {
+			vm, err := tests.NewRandomVMWithUserData("noCloud", userData)
+			Expect(err).ToNot(HaveOccurred())
+			obj := LaunchVM(vm)
+			VerifyUserDataVM(vm, obj)
```
Comment (rmohr) [inline tests/vm_userdata_test.go]: Could you pass in the magic string to this function? It's quite hard to follow against what you are verifying otherwise.

Comment (davidvossel) [reply tests/vm_userdata_test.go]: @rmohr userData and magicStr and now something that are generated in the same place the vm is created, launched, and verified from.   I agree this makes the tests a little easier to follow.

**Comment by rmohr** [comment]
@kubevirt-bot retest this please

**Comment by rmohr** [comment]
retest this please
